### PR TITLE
[feat] Reserve Auction v1.0

### PR DESCRIPTION
**Pull request by kulkarohan**
# New Module Proposal

## Description
This module outlines a way for any ERC-721 token holder to create a permissionless reserve auction.

## Motivation and Context
ReserveAuctionV1 is an update to ZORA V2 [Auction House](https://github.com/ourzora/auction-house/). NFTs are taken into escrow after the reserve price is met, as well as the current highest bid. After an auction has ended, anyone can settle to payout respective parties and transfer the NFT to the winning bidder.

## How has this been tested?

Relevant tests have been included via Foundry.

## Checklist:

- [x] The module includes tests written for Foundry
- [x] The module is documented with [NATSPEC](https://docs.soliditylang.org/en/v0.5.10/natspec-format.html)
- [x] The documentation includes [UML Diagrams](https://plantuml.com/ascii-art) for external and public functions
- [x] The module is a [Hyperstructure](https://www.jacob.energy/hyperstructures.html)

### CODE REVIEW

**Review by almndbtr — COMMENTED** [review]
@ourzora - Thanks again for opening this to community review. ✌🏽

File: contracts/test/modules/ReserveAuction/V1/ReserveAuction.integration.t.sol [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+        vm.warp(1 days);
+        vm.prank(address(bidder));
+        auctions.createBid{value: 1 ether}(address(token), 0, 1 ether, address(finder));
+
+        vm.warp(1 days + 1 hours);
+        auctions.settleAuction(address(token), 0);
+    }
+
+    function test_ETHIntegration() public {
```
Comment (almndbtr) [inline contracts/test/modules/ReserveAuction/V1/ReserveAuction.integration.t.sol:136]: The assertions in this function take into account the royalties paid out to the creator and the finder respectively. However, there doesn’t seem to be an assertion that checks the balances if a protocol fee is set at the time the auction is settled:

https://github.com/ourzora/v3/blob/a5b3c77964c48de5d0cb16a33b2efe135662459d/contracts/modules/ReserveAuction/V1/ReserveAuctionV1.sol#L494

Even if no protocol fee were actually set in these test cases (zero), would the team be open to adding one more require to show that is truly the case, or include a separate set of tests that illustrate what happens when an auction is settled with a non-zero protocol fee? 💭

**Comment by almndbtr** [comment]
👋 @kulkarohan -- I see some updates being pushed 👀 

Is it best to hold off from further review until those settle? 💭

**Comment by kulkarohan** [comment]
hey @almndbtr sorry about that haha we found a few small tweaks earlier that we wanted to make. feel free to further review!

File: contracts/modules/ReserveAuction/V1/ReserveAuctionV1.sol [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+
+        // If first bid --
+        bool firstBid;
+        if (auction.firstBidTime == 0) {
+            // Store time of bid
+            auction.firstBidTime = block.timestamp;
+            // Mark as first bid
+            firstBid = true;
+            // Transfer NFT into escrow
+            erc721TransferHelper.transferFrom(_tokenContract, auction.seller, address(this), _tokenId);
```
[the remark this answers is not in the document]
Comment (kulkarohan) [reply contracts/modules/ReserveAuction/V1/ReserveAuctionV1.sol:396]: good question - this is by design. `safeTransferFrom` is protection on the user side to prevent sending NFTs to contracts that may not be able to handle them. in our case, since we know we're the recipient and can handle them, we can use the simpler `transferFrom` and save gas

File: contracts/modules/ReserveAuction/V1/ReserveAuctionV1.sol [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+import {IncomingTransferSupportV1} from "../../../common/IncomingTransferSupport/V1/IncomingTransferSupportV1.sol";
+import {FeePayoutSupportV1} from "../../../common/FeePayoutSupport/FeePayoutSupportV1.sol";
+import {ModuleNamingSupportV1} from "../../../common/ModuleNamingSupport/ModuleNamingSupportV1.sol";
+
+/// @title Reserve Auction V1
+/// @author tbtstl <t@zora.co>
+/// @notice This contract allows users to list and bid on ERC-721 tokens with timed reserve auctions
+contract ReserveAuctionV1 is ReentrancyGuard, UniversalExchangeEventV1, IncomingTransferSupportV1, FeePayoutSupportV1, ModuleNamingSupportV1 {
+    /// @dev The indicator to pass all remaining gas when paying out royalties
+    uint256 private constant USE_ALL_GAS_FLAG = 0;
```
Comment (tbtstl) [inline contracts/modules/ReserveAuction/V1/ReserveAuctionV1.sol:18]: unused var here

File: contracts/modules/ReserveAuction/V1/ReserveAuctionV1.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (almndbtr) [reply contracts/modules/ReserveAuction/V1/ReserveAuctionV1.sol:396]: Ah, that context helps! Thank you @kulkarohan.

I was mulling over proposing a follow up change including an annotation there, but I think your input is enough for someone to look at this in the future and figure out why we use the simpler function. Considering this resolved 👍 

```suggestion
            // `transferFrom` is enough to move the NFT into escrow since the contract is the recipient
            erc721TransferHelper.transferFrom(_tokenContract, auction.seller, address(this), _tokenId);
```

File: contracts/modules/ReserveAuction/V1/ReserveAuctionV1.sol (hunk 3 above)
Comment (almndbtr) [reply contracts/modules/ReserveAuction/V1/ReserveAuctionV1.sol:18]: I cut a PR to replace this with a new constant for the outgoing transfer gas limit: https://github.com/ourzora/v3/pull/137

File: contracts/test/modules/ReserveAuction/V1/ReserveAuction.integration.t.sol (hunk 1 above)
Comment (almndbtr) [reply contracts/test/modules/ReserveAuction/V1/ReserveAuction.integration.t.sol:136]: I see your 👍 @kulkarohan -- I'll take some time this week and cut a PR that introduces a test for the protocol fee. 🔪

(If there's a known, hard deadline, do let me know and I can prioritize accordingly.)

Comment (almndbtr) [reply contracts/test/modules/ReserveAuction/V1/ReserveAuction.integration.t.sol:136]: @kulkarohan - Hi again! 👋

I opened https://github.com/ourzora/v3/pull/140 introducing tests for when the protocol fee is set to 1 basis point (`1 bps` or `0.01%`). I'd love your evaluative review on it so we have complete test coverage for this module. Thanks in advance!

**Comment by reefloretto** [comment]
hey! are there outstanding tasks required for this to be considered deploy-ready? my team is working on auction support for media tokens interfacing w/ the zora V3 asks module & it would be amazing to support V3 reserve auctions as well.

thank you for pushing the space forward. ⚡

**Comment by kulkarohan** [comment]
hey @reefloretto ! we're just restructuring a couple things that we'll push today. planning to deploy Friday!

**Comment by almndbtr** [comment]
> planning to deploy Friday!

🙏 Hoping the deploy goes well! Godspeed, Zora team 🚀

File: contracts/modules/ReserveAuction/V1/ReserveAuctionV1.sol [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+    /// @param startTime The start time of the auction
+    /// @param firstBidTime The time of the first bid
+    /// @param reservePrice The price to create the first bid
+    struct Auction {
+        address seller;
+        address currency;
+        address sellerFundsRecipient;
+        address bidder;
+        address finder;
+        uint16 findersFeeBps;
```
Comment (iainnash) [inline contracts/modules/ReserveAuction/V1/ReserveAuctionV1.sol:49]: One way to save 2 slots would be to limit the size of a few fields to a more reasonable number.

Duration can easily be `uint32` (probably not a great idea to have an auction beyond 100 years etc).

`uint32+uint16+address` is still less than `uint256` (one storage slot)

`firstBidTime` and `startTime` can both be `uint126` to save one slot as well

```
address
uint16 finderFeeBps
uint32 duration
uint128 firstBidTime
uint128 startTime
```

saves two slots over the current configuration.

Comment (kulkarohan) [reply contracts/modules/ReserveAuction/V1/ReserveAuctionV1.sol:49]: @iainnash hahaha i have these changes in the update. pushing soon!

**Comment by kulkarohan** [comment]
@everyone

I've restructured Reserve Auction into four gas-optimized variants:
- Core ETH
- Core ERC-20
- Finders ETH
- Finders ERC-20

'Core' are minimal reserve auctions and 'Finders' add support for finders fee.

An important note: while ETH variants are ETH-only, the ERC-20 variants are still compatible with ETH via setting `address(0)` as the currency.

So if you'd like the simplicity of interacting with a sole contract, the Finders ERC-20 variant will support all four use-cases (ETH, ERC20s, no finders fee, finders fee) as initially designed. But if you'd like to optimize per use-case, you'll now be able to do so as well.

Lemme know if there are any Qs -- thanks for your patience.
